feat: Follow consistent script conventions

espendk opened the pull request:
The following conventions have been applied to all *.s.sol scripts:

1. Env vars are read in the `run` function, not the `innerRun` function.
   This ensures that env vars are only read by the script the user invoked and will not accidentally be picked up (and possibly misinterpreted) by another script.

3. Env vars specifying contracts can always contain either a contract name or an address.
   This functionality (provided by the envAddressOrName function) was used in some places, but not all, resulting in an inconsistent user experience. Contract names are resolved via the `fork.get` function, ie the ToyENS registry.

5. Contract names are resolved in the `run` function, not the `innerRun` function.
   This helps ensure that contract addresses specified by the user when calling the outermost script are not ignored by another script (which could happen before because scripts relied on the address provided by fork.get for a hardcoded name).

6. Contract addresses can always be specified by the user via env vars.
   Instead of always of relying on all addresses being registered in ToyENS and looking these up via hardcoded names, all scripts now allow contract addresses/names to be specified in env vars.

7. Specific contract/interface types are preferred for `innerRun` parameters instead of `address`.
   This reduces the risk of passing a wrong a address from another script.

8. `outputDeployment` is called in the `run` function, not in `innerRun`.
   This ensures the outermost script is in control of when the deployment is output.

espendk committed:
feat: Make scripts consistent

The following conventions have been applied to all *.s.sol scripts:

1. Env vars are read in the `run` function, not the `innerRun` function.
This ensures that env vars are only read by the script the user invoked and will not accidentally be picked up (and possibly misinterpreted) by another script.

2. Env vars specifying contracts can always contain either a contract name or an address.
This functionality (provided by the envAddressOrName function) was used in some places, but not all, resulting in an inconsistent user experience.
Contract names are resolved via the `fork.get` function, ie the ToyENS registry.

3. Contract names are resolved in the `run` function, not the `innerRun` function.
This helps ensure that contract addresses specified by the user when calling the outermost script are not ignored by another script (which could happen before because scripts relied on the address provided by fork.get for a hardcoded name).

4. Contract addresses can always be specified by the user via env vars.
Instead of always of relying on all addresses being registered in ToyENS and looking these up via hardcoded names, all scripts now allow contract addresses/names to be specified in env vars.

5. Specific contract/interface types are preferred for `innerRun` parameters instead of `address`.
This reduces the risk of passing a wrong a address from another script.

6. `outputDeployment` is called in the `run` function, not in `innerRun`
This ensures the outermost script is in control of when the deployment is output.

peterMangrove commented on script/ActivateMarket.s.sol:
```diff
@@ -19,12 +19,13 @@ import {ActivateSemibook} from "./ActivateSemibook.s.sol";
   function run() public {
     innerRun({
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      reader: MgvReader(envHas("MGV_READER") ? envAddressOrName("MGV_READER") : fork.get("MgvReader")),
```
I see this pattern being used in a lot of scripts. Maybe we should create a helper function in `Deployer`, that can do this, in order to insure the patteren.

espendk replied:
Good idea, implemented

peterMangrove commented on script/ActivateSemibook.s.sol:
```diff
@@ -25,16 +25,15 @@ uint constant COVER_FACTOR = 1000;
     innerRun({
-      outbound_tkn: vm.envAddress("OUTBOUND_TKN"),
-      inbound_tkn: vm.envAddress("INBOUND_TKN"),
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
```
Same pattern

peterMangrove commented on script/DeactivateMarket.s.sol:
```diff
@@ -9,21 +9,23 @@ import {UpdateMarket} from "mgv_script/periphery/UpdateMarket.s.sol";
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      reader: MgvReader(envHas("MGV_READER") ? envAddressOrName("MGV_READER") : fork.get("MgvReader")),
+      tkn0: envAddressOrName("TKN0"),
+      tkn1: envAddressOrName("TKN1")
```
Should maybe be cast to `IERC20`

espendk replied:
Agreed, implemented everywhere :-)

peterMangrove commented on script/ActivateSemibook.s.sol:
```diff
@@ -25,16 +25,15 @@ uint constant COVER_FACTOR = 1000;
-      inbound_tkn: vm.envAddress("INBOUND_TKN"),
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      outbound_tkn: envAddressOrName("OUTBOUND_TKN"),
+      inbound_tkn: envAddressOrName("INBOUND_TKN"),
```
`IERC20`?

peterMangrove commented on script/ActivateMarket.s.sol:
```diff
@@ -19,12 +19,13 @@ import {ActivateSemibook} from "./ActivateSemibook.s.sol";
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      reader: MgvReader(envHas("MGV_READER") ? envAddressOrName("MGV_READER") : fork.get("MgvReader")),
       tkn1: envAddressOrName("TKN1"),
       tkn2: envAddressOrName("TKN2"),
```
`IERC20`?

peterMangrove commented on script/UseOracle.s.sol:
```diff
@@ -7,15 +7,13 @@ import {Mangrove} from "mgv_src/Mangrove.sol";
-      oracleAddress: envHas("ORACLE") ? vm.envAddress("ORACLE") : fork.get("MgvOracle"),
-      mgvAddress: payable(envHas("MGV") ? vm.envAddress("MGV") : fork.get("Mangrove"))
+      mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      oracleAddress: envHas("ORACLE") ? envAddressOrName("ORACLE") : fork.get("MgvOracle")
```
Maybe cast to `MgvOracle`?

peterMangrove commented on script/periphery/UpdateMarket.s.sol:
```diff
@@ -13,30 +13,26 @@ import "forge-std/console.sol";
       tkn0: envAddressOrName("TKN0"),
-      tkn1: envAddressOrName("TKN1"),
-      mgvReaderAddress: payable(envHas("MGV_READER") ? vm.envAddress("MGV_READER") : fork.get("MgvReader"))
+      tkn1: envAddressOrName("TKN1")
```
`IERC20`?

peterMangrove commented on script/strategies/kandel/KandelDeployer.s.sol:
```diff
@@ -20,12 +20,14 @@ contract KandelDeployer is Deployer {
     innerRun({
+      mgv: IMangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
       base: envAddressOrName("BASE"),
       quote: envAddressOrName("QUOTE"),
```
`IERC20`?

peterMangrove commented on script/strategies/kandel/KandelSower.s.sol:
```diff
@@ -18,27 +18,44 @@ import {MangroveTest, Test} from "mgv_test/lib/MangroveTest.sol";
+    bool onAave = vm.envBool("ON_AAVE");
     innerRun({
+      mgv: IMangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
+      kandelSeeder: envHas("KANDEL_SEEDER")
```
Maybe cast to `AbstractKandelSeeder`

peterMangrove commented on script/strategies/mango/MangoDeployer.s.sol:
```diff
@@ -32,17 +32,21 @@ contract MangoDeployer is Deployer {
     innerRun({
+      mgv: IMangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove")),
       base: envAddressOrName("BASE"),
       quote: envAddressOrName("QUOTE"),
```
`IERC20`?

peterMangrove commented on script/strategies/mangroveOrder/ActivateMangroveOrder.s.sol:
```diff
@@ -12,26 +12,35 @@ import {Deployer} from "mgv_script/lib/Deployer.sol";
   function run() public {
-    innerRun({mgvOrder: MangroveOrder(fork.get("MangroveOrder")), tkns: vm.envString("TKNS", ",")});
+    string[] memory tkns_env = vm.envString("TKNS", ",");
+    address[] memory tkns = new address[](tkns_env.length);
```
Maybe array of `IERC20` instead of `address`?

peterMangrove commented on script/toy/AmplifierDeployer.s.sol:
```diff
@@ -14,10 +14,11 @@ import {Deployer} from "mgv_script/lib/Deployer.sol";
+      admin: envAddressOrName("ADMIN"),
+      base: envHas("BASE") ? envAddressOrName("BASE") : fork.get("WETH"),
+      stable1: envHas("STABLE1") ? envAddressOrName("STABLE1") : fork.get("USDC"),
+      stable2: envHas("STABLE2") ? envAddressOrName("STABLE2") : fork.get("DAI")
```
`IERC20`?

espendk committed:
feat: Add overload of envAddressOrName that takes a default address

Simplifies this common pattern:

```
    innerRun({mgv: Mangrove(envHas("MGV") ? envAddressOrName("MGV") : fork.get("Mangrove"))});
```

to

```
    innerRun({mgv: Mangrove(envAddressOrName("MGV", fork.get("Mangrove")))});
```

espendk committed:
refactor: Specific types instead of address whenever possible in innerRun

espendk committed:
docs: Add README.md to scripts folder

espendk commented:
Thanks for the input @peterMangrove 🙏 I've implemented your suggestions and would appreciate a new review :-)

peterMangrove commented on script/strategies/mango/MangoDeployer.s.sol:
```diff
@@ -32,51 +32,62 @@ contract MangoDeployer is Deployer {
       price_incr: vm.envUint("PRICE_INCR"),
-      admin: vm.envAddress("DEPLOYER")
+      admin: envAddressOrName("DEPLOYER"),
+      name: envHas("NAME") ? vm.envString("NAME") : ""
```
Should we have a helper function in `Deployer`, just like `envAddressOrName` ?

espendk replied:
hmm, not sure - the Mango family is the only one using this pattern and I'm unsure if we want to promote it.
Let's leave it for now.
